### ISSUE TITLE
NU1608 warning and runtime error when adding ReportViewerCore to a .NET 8 web project that uses Razor Runtime Compilation

### ISSUE THREAD

**Issue by pmarflee** [lkosson/reportviewercore#162]
Steps to reproduce:

1. Create a new .NET 8 ASP.NET Core Web App (Razor Pages) project
2. Add the following package references to the project file:

```
  <ItemGroup>
    <PackageReference Include="ReportViewerCore.NETCore" Version="15.1.19" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="8.0.0" />
  </ItemGroup>
```

3. Add the following code to Program.cs:

```
var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services
    .AddRazorPages()
    .AddRazorRuntimeCompilation();

var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment())
{
    app.UseExceptionHandler("/Error");
}
app.UseStaticFiles();

app.UseRouting();

app.UseAuthorization();

app.MapRazorPages();

app.Run();
```

At this point, the following warning will be displayed in the 'Error List' window:

> warning NU1608: Detected package version outside of dependency constraint: Microsoft.CodeAnalysis.CSharp 4.0.0 requires Microsoft.CodeAnalysis.Common (= 4.0.0) but version Microsoft.CodeAnalysis.Common 4.8.0-3.final was resolved

4. Press F5 to start the website. The following exception will be thrown on line 21: `app.MapRazorPages()`

> Declaration referenced in a method implementation cannot be a final method.  Type: 'Microsoft.CodeAnalysis.CSharp.CSharpCompilationOptions'.  Assembly: 'Microsoft.CodeAnalysis.CSharp, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.

I can get rid of the runtime error (but not the NUS1608 warning) by updating my package references to the following:

```
  <ItemGroup>
    <PackageReference Include="ReportViewerCore.NETCore" Version="15.1.19" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="8.0.0" />
    <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.8.0" />
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.8.0" />
  </ItemGroup>
```

Is this an issue with ReportViewerCore? Or the Microsoft Razor Runtime Compilation library?

**Comment by lkosson** [comment]
As far as I can tell, this is just a deficiency in NuGet dependencies resolution mechanism. `Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation` version `8.0.0` calls for `Microsoft.CodeAnalysis.Razor` version `6.0.0` or higher, which in turn depends `Microsoft.CodeAnalysis.CSharp` version `4.0.0` or higher which requires `Microsoft.CodeAnalysis.Common` of exact version `4.0.0`. On the other hand `ReportViewerCore.NETCore` for .NET 8, due to convoluted reasons, requires `Microsoft.CodeAnalysis.VisualBasic` of exact version `4.8.0-3.final`, which in turn requires exact version `4.8.0-3.final` of `Microsoft.CodeAnalysis.Common`.

In reality, both `Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation` and `ReportViewerCore.NETCore` can live with almost any version of `Microsoft.CodeAnalysis.Common`, as long as it is same version as `Microsoft.CodeAnalysis.CSharp` and `Microsoft.CodeAnalysis.VisualBasic`. However, for some unknown reasons NuGet resolves `Microsoft.CodeAnalysis.CSharp` to version `4.0.0` and `Microsoft.CodeAnalysis.Common` to version `4.8.0`.

Your solution of adding explicit reference to `Microsoft.CodeAnalysis.*` packages should be fine. In fact, only adding `Microsoft.CodeAnalysis.CSharp` reference is enough in this case.

**Comment by pmarflee** [comment]
Thanks for your response. Microsoft.CodeAnalysis.VisualBasic is now at version 4.8.0, published 7 days ago. Does ReportViewerCore Nuget package need an update to target this newer version?

**Comment by lkosson** [comment]
Not really. `Microsoft.VisualStudio.Web.CodeGeneration`, which gets installed when you use "Add scaffolded item" in VS, still references `Microsoft.DotNet.Scaffolding.Shared` version `8.0.0`, which references version `4.8.0-3.final` of `Microsoft.CodeAnalysis.CSharp.Features`, requiring exact matching version of `Microsoft.CodeAnalysis.Common`.

**Comment by pmarflee** [comment]
OK, I'll add the `Microsoft.CodeAnalysis.*` packages to my solution as a workaround to this problem. Thanks for your help.
